src/main/java/com/microsoft/sqlserver/jdbc/KeyVaultCustomCredentialPolicy.java

```diff
@@ -0,0 +1,117 @@
+/*
+ * Microsoft JDBC Driver for SQL Server Copyright(c) Microsoft Corporation All rights reserved. This program is made
+ * available under the terms of the MIT License. See the LICENSE file in the project root for more information.
+ */
+
+package com.microsoft.sqlserver.jdbc;
+
+import com.azure.core.credential.TokenRequestContext;
+import com.azure.core.http.HttpPipelineCallContext;
+import com.azure.core.http.HttpPipelineNextPolicy;
+import com.azure.core.http.HttpResponse;
+import com.azure.core.http.policy.HttpPipelinePolicy;
+import com.azure.core.util.CoreUtils;
+
+import java.text.MessageFormat;
+import java.util.HashMap;
+import java.util.Locale;
+import java.util.Map;
+import reactor.core.publisher.Mono;
+
+/**
+ * A policy that authenticates requests with Azure Key Vault service.
+ */
+class KeyVaultCustomCredentialPolicy implements HttpPipelinePolicy {
+    private static final String WWW_AUTHENTICATE = "WWW-Authenticate";
+    private static final String BEARER_TOKEN_PREFIX = "Bearer ";
+    private static final String AUTHORIZATION = "Authorization";
+    private final ScopeTokenCache cache;
+    private final KeyVaultTokenCredential keyVaultTokenCredential;
+
+    /**
+     * Creates KeyVaultCustomCredentialPolicy.
+     *
+     * @param credential
+     *        the token credential to authenticate the request
+     * @throws SQLServerException
+     */
+    KeyVaultCustomCredentialPolicy(KeyVaultTokenCredential credential) throws SQLServerException {
+        if (null == credential) {
+            MessageFormat form = new MessageFormat(SQLServerException.getErrString("R_NullValue"));
+            Object[] msgArgs1 = {"Credential"};
+            throw new SQLServerException(form.format(msgArgs1), null);
+        }
+
+        this.cache = new ScopeTokenCache(credential::getToken);
+        this.keyVaultTokenCredential = credential;
+    }
+
+    /**
+     * Adds the required header to authenticate a request to Azure Key Vault service.
+     *
+     * @param context
+     *        The request context
+     * @param next
+     *        The next HTTP pipeline policy to process the {@code context's} request after this policy completes.
+     * @return A {@link Mono} representing the HTTP response that will arrive asynchronously.
+     */
+    @Override
+    public Mono<HttpResponse> process(HttpPipelineCallContext context, HttpPipelineNextPolicy next) {
+        if ("http".equals(context.getHttpRequest().getUrl().getProtocol())) {
```

Comment:
If only HTTPS supported lets change this to  `if (!"https".equals(context.getHttpRequest().getUrl().getProtocol())) {`